component-tests/steps/authoriseSteps.ts

```diff
@@ -0,0 +1,227 @@
+import { Given, When, Then, After } from 'cucumber';
+import { Mock, It, Times, IMock } from 'typemoq';
+import { expect } from 'chai';
+import * as sinon from 'sinon';
+import * as crypto from 'crypto';
+import { Buffer } from 'buffer';
+import { CustomAuthorizerEvent, CustomAuthorizerResult } from 'aws-lambda';
+import * as jsonwebtoken from 'jsonwebtoken';
+import * as testKeys from './testKeys';
+import * as authoriser from '../../src/functions/authoriser/framework/handler';
+import AdJwtVerifier, { JwksClient } from
+  '../../src/functions/authoriser/application/AdJwtVerifier';
+
+const uuid = () => crypto.randomBytes(16).toString('hex');
+const oneMinute = 60;
+
+const createToken = (
+  context: AuthoriseStepsContext,
+  notBefore?: number,
+  expiresIn?: number,
+  privateKey?: string,
+  issuer?: string,
+  audience?: string): string => {
+  const payload = {
+    unique_name: context.testTokenUniqueName,
+  };
+
+  const signOptions: jsonwebtoken.SignOptions = {
+    algorithm: 'RS256',
+    keyid: context.testKid,
+    audience: audience || context.testAppId,
+    issuer: issuer || context.testIssuer,
+    subject: context.testTokenSubject,
+    notBefore: notBefore || -oneMinute,
+    expiresIn: expiresIn || (oneMinute * 5),
+  };
+
+  return jsonwebtoken.sign(payload, privateKey || testKeys.ourCertificate.privateKey, signOptions);
+};
+
+const base64Decode = (base64String: string): { [propName: string]: any } => {
+  const buffer = new Buffer(base64String, 'base64');
+  const json = buffer.toString('ascii');
+  const object = JSON.parse(json);
+  return object;
+};
+
+const base64Encode = (object: any): string => {
+  const json = JSON.stringify(object);
+  const buffer = new Buffer(json);
+  const base64String = buffer.toString('base64');
+  return base64String;
+};
+
+interface AuthoriseStepsContext {
+  sinonSandbox: sinon.SinonSandbox;
+  sut: (event: CustomAuthorizerEvent) => Promise<CustomAuthorizerResult>;
+  moqConsoleLog: IMock<(message?: any, ...optionalParams: any[]) => void>;
+  moqJwksClient: IMock<JwksClient>;
+  testAppId: string;
+  testIssuer: string;
+  testKid: string;
+  testTokenUniqueName: string;
+  testTokenSubject: string;
+  token: string;
+  methodArn: string;
+  result?: CustomAuthorizerResult;
+}
+
+After(function () {
+  const context: AuthoriseStepsContext = this.context;
```

Comment:
Don't think the use of `this` causes a problem but I think it will be referring at the global scope in this instance? Do we want to just create a `let context;` at the top and then operate on that?

Reply:
Hi @ammarhaiderbjss, its a good question.  The `this` object inside `cucumber-js` steps is unique per scenario, and I can confirm that after looking this up, the `cucumber-js` docs say that they have a concept called a "World", which is an isolated context for each scenario, exposed to the hooks and steps as `this`.  I realise the use of `this` is not intuitive to actually what is being referred to, but its just the way `cucumber-js` works.  Essentially it means inside any cucumber steps, `this` refers to a unique object whose purpose is to share data between steps.  The problem with having a context in the shared scope is it would essentially mean the context is shared between scenarios.